### PR TITLE
ARROW-14656: [Python] sort helper for Array, ChunkedArray and StructArray

### PR DESCRIPTION
**Pull request by amol-**

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -2360,6 +2360,25 @@ cdef class StructArray(Array):
         result.validate()
         return result
 
+    def sort_by(self, fieldname, order="ascending"):
```
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:2363]: It seems a bit strange to have this method only on StructArray IMO. Also for other arrays, sorting takes some effort with the current API.

Comment (pitrou) [reply python/pyarrow/array.pxi:2363]: This is sorting by a given child field. So it wouldn't make sense on other arrays (including unions).

Comment (jorisvandenbossche) [reply python/pyarrow/array.pxi:2363]: But it does make sense to sort just the array. 

What I mean is, also for sorting an array, it takes some code:

```
arr.take(pc.array_sort_indices(arr, order))
```

vs for struct array:

```
arr.take(pc.array_sort_indices(arr.field(name), order)
```

Once you have the first snippet, I don't see why the second snippet would necessarily warrant a helper function (and the first not)

Comment (pitrou) [reply python/pyarrow/array.pxi:2363]: Well, we can expose both if we want, though @amol- 's ticket seems to be specifically for `sort_by` :-)

Comment (jorisvandenbossche) [reply python/pyarrow/array.pxi:2363]: I know, but so my feedback is that I am personally -1 on adding a special case method `sort_by`

**Comment by amol-** [comment]
This is currently blocked by:
* `array_sort_indices` not working on ChunkedArray: https://issues.apache.org/jira/browse/ARROW-12042
* `array_sort_indices` not implemented for StructArray

**Comment by pitrou** [comment]
@amol- Why don't you just `sort_indices` instead?

**Comment by amol-** [comment]
> @amol- Why don't you just `sort_indices` instead?

That's what I'm ending up doing for `ChunkedArray`, I tried passing `None` as the column name but it explicitly requires a string, so I'll just end up passing an empty one

**Comment by jorisvandenbossche** [comment]
> That's what I'm ending up doing for `ChunkedArray`, I tried passing `None` as the column name but it explicitly requires a string, so I'll just end up passing an empty one

Yes, using "dummy" name is also what we do elsewhere:

https://github.com/apache/arrow/blob/92ee2952155070f7c0bfe3c05f9ccddebf7620db/python/pyarrow/compute.py#L722-L723

File: python/pyarrow/tests/test_array.py [hunk 2]
```diff
@@ -732,19 +732,19 @@ def test_struct_array_sort():
         {"a": 5, "b": "foo"},
     ]
 
-    sorted_arr = arr.sort("descending")
+    sorted_arr = arr.sort("ascending", fieldname="b")
```
Comment (amol-) [inline python/pyarrow/tests/test_array.py:735]: This should be revert to omit the `fieldname`, it was the acceeptance test for sorting struct arrays wiithout any specific field.

File: python/pyarrow/tests/test_array.py [hunk 3]
```diff
@@ -732,19 +732,19 @@ def test_struct_array_sort():
...
-        {"a": 35, "b": "foobar"},
-        {"a": 7, "b": "car"},
         {"a": 7, "b": "bar"},
+        {"a": 7, "b": "car"},
         {"a": 5, "b": "foo"},
+        {"a": 35, "b": "foobar"},
     ]
 
-    sorted_arr = arr.sort("ascending")
+    sorted_arr = arr.sort("ascending", fieldname="a")
```
Comment (amol-) [inline python/pyarrow/tests/test_array.py:743]: This should be revert to omit the `fieldname`, it was the acceeptance test for sorting struct arrays wiithout any specific field.

**Comment by amol-** [comment]
> @amol- What is the status on this? Do you want to take it up?

I think I can take it up if @Jedi18 can't work on it anymore. Or even better would @AlvinJ15 be able to tackle it? He can probably do it faster than me.

**Comment by jorisvandenbossche** [comment]
@amol- I think this PR can be closed?

The python API additions have been merged in https://github.com/apache/arrow/pull/14781, and the C++ additions are also included in the open PR https://github.com/apache/arrow/pull/14369

**Comment by amol-** [comment]
> @amol- I think this PR can be closed?
> 
> The python API additions have been merged in #14781, and the C++ additions are also included in the open PR #14369

I think so, I think it was still open only to take the pieces and move them into the other PRs

**Pull request closed without merging**